src/Shared/FileMatcher.cs

```diff
@@ -1905,21 +2123,42 @@ DirectoryExists directoryExists
              * This is because it's cheaper to add items to an IList and this code
              * might potentially do a lot of that.
              */
-            var listOfFiles = new List<string>();
+            var listOfFiles = new ConcurrentStack<string>();
 
             /*
              * Now get the files that match, starting at the lowest fixed directory.
              */
             try
             {
+                // Setup the values for calculating the MaxDegreeOfParallelism option of Parallel.ForEach
+                // Set to use only half processors when we have 4 or more of them, in order to not be too aggresive
+                // By setting MaxTasksPerIteration to the maximum amount of tasks, which means that only one
+                // Parallel.ForEach will run at once, we get a stable number of threads being created.
+                var maxTasks = Math.Max(1, Environment.ProcessorCount / 2);
+                var taskOptions = new TaskOptions(maxTasks)
+                {
+                    AvailableTasks = maxTasks,
+                    MaxTasksPerIteration = maxTasks
+                };
                 GetFilesRecursive(
                     listOfFiles,
                     state,
                     projectDirectoryUnescaped,
                     stripProjectDirectory,
                     getFileSystemEntries,
                     searchesToExclude,
-                    searchesToExcludeInSubdirs);
+                    searchesToExcludeInSubdirs,
+                    taskOptions);
+            }
+            // Catch exceptions that are thrown inside the Parallel.ForEach
+            catch (AggregateException ex)
+            {
+                // Flatten to get exceptions than are thrown inside a nested Parallel.ForEach
+                if (ex.Flatten().InnerExceptions.All(ExceptionHandling.IsIoRelatedException))
+                {
+                    return CreateArrayWithSingleItemIfNotExcluded(filespecUnescaped, excludeSpecsUnescaped);
```

Comment:
Oh, is this to handle the exceptions? If so, why we do we still need the catch blocks above?

Reply:
Check the [comment](https://github.com/Microsoft/msbuild/pull/2572#discussion_r143182574) above for catching the `AggregateException`.